performant-software/dm-2
# Add canvas documents with image uploading

### What does this PR do?
- Adds features for the creation and editing of canvas (image) documents
- Includes support for image uploading, as well as IIIF info.json and external static image URLs, for canvas tile sources
- Provides drawing tools for adding highlights to a canvas document
- Generates thumbnails for canvas documents and their highlights

### How to test
- In the [review app](https://dm-2-staging-pr-38.herokuapp.com/), open a new or existing project and create a new image document from the 'Add new document' button in the sidebar.
- The canvas document should open in the grid in its 'Add image source' mode - here you can select the type of image source to add using a dropdown menu, paste in the URL or select a file to upload, and confirm the addition of the image source.
- Upload an image to use as the document's base layer by selecting 'Upload image' in the dropdown, then clicking 'Choose files'. After selecting an image and hitting OK, you should see a progress indicator and then a small thumbnail of the image when uploaded. Click 'Add image source' to complete the addition of the uploaded image. 
- The document should switch to its normal viewer mode after you click 'Add image source', and after a pause, the image should appear both in its large form in the zoomable document viewer and as a thumbnail for the document in the table of contents. Try panning and zooming in the image viewer, using the slider to the left of the image to zoom as well as scrolling and the +/- buttons. The slider's position should be affected by the other methods of zooming.
- Add highlights using the various drawing tools. The rectangle, circle, and marker tools should add objects directly to the canvas which can then be moved and resized (markers can't be resized). The free-draw and line-draw buttons should toggle their respective modes on and off, changing the behavior of clicking/dragging on the image. Once these modes are exited, the objects created should be selectable and changeable as with the other shapes.
- Hovering over a highlight should bring up its link inspector popup. You should see a thumbnail for the highlight that looks (roughly) like the shape overlaid on its area of the image, but these may take a few seconds to generate after highlights are created/changed.
- Delete highlights using the delete icon next to the drawing tools. A confirmation dialog should appear before the highlight is destroyed.
